src/community/schemaless-features/schemaless-core/src/main/java/org/geoserver/schemalessfeatures/data/SchemalessFeatureCollection.java

```diff
@@ -0,0 +1,98 @@
+/* (c) 2021 Open Source Geospatial Foundation - all rights reserved
+ * This code is licensed under the GPL 2.0 license, available at the root
+ * application directory.
+ */
+package org.geoserver.schemalessfeatures.data;
+
+import java.io.IOException;
+import java.util.Collection;
+import org.geotools.data.Query;
+import org.geotools.feature.FeatureCollection;
+import org.geotools.feature.FeatureIterator;
+import org.geotools.geometry.jts.ReferencedEnvelope;
+import org.opengis.feature.Feature;
+import org.opengis.feature.FeatureVisitor;
+import org.opengis.feature.type.FeatureType;
+import org.opengis.filter.Filter;
+import org.opengis.filter.sort.SortBy;
+import org.opengis.util.ProgressListener;
+
+public class SchemalessFeatureCollection implements FeatureCollection<FeatureType, Feature> {
+
+    protected SchemalessFeatureSource featureSource;
+
+    protected Query query;
+
+    public SchemalessFeatureCollection(Query q, SchemalessFeatureSource source) {
+        this.query = q;
+        this.featureSource = source;
+    }
+
+    @Override
+    public FeatureIterator<Feature> features() {
+        return new SchemalessFeatureIterator(featureSource.getReader(this.query));
+    }
+
+    @Override
+    public FeatureType getSchema() {
+        return featureSource.getSchema();
+    }
+
+    @Override
+    public String getID() {
+        return null;
+    }
+
+    @Override
+    public void accepts(FeatureVisitor visitor, ProgressListener progress) throws IOException {}
+
+    @Override
+    public FeatureCollection<FeatureType, Feature> subCollection(Filter filter) {
+        return null;
```

Comment:
pretty easy to add since this class takes a query, see https://github.com/geotools/geotools/blob/f66689b5305eaf381e21dc2feec8011138d4cb24/modules/library/main/src/main/java/org/geotools/data/store/ContentFeatureCollection.java#L342